core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -82,12 +108,28 @@ public CacheEventNotificationServiceImpl(ExecutorService orderedDelivery, Execut
   public void registerCacheEventListener(CacheEventListener<? super K, ? super V> listener,
                                   EventOrdering ordering, EventFiring firing, EnumSet<EventType> forEventTypes) {
     boolean doRegister = forEventTypes.contains(EventType.EVICTED) || forEventTypes.contains(EventType.EXPIRED);
-    if (!registeredListeners.add(new EventListenerWrapper(listener, firing, ordering, forEventTypes))) {
+    EventListenerWrapper wrapper = new EventListenerWrapper(listener, firing, ordering, forEventTypes);
+    if(wrapper.config.orderingMode() == EventOrdering.ORDERED) {
+      orderedListenerCount.incrementAndGet();
+    }
+
+    if(aSyncListenersSet.contains(wrapper) || syncListenersSet.contains(wrapper)) {
       throw new IllegalStateException("Cache Event Listener already registered: " + listener);
     }
+
+    if (firing == EventFiring.ASYNCHRONOUS) {
+      aSyncListenersSet.add(wrapper);
+    } else {
+      syncListenersSet.add(wrapper);
+    }
     if (doRegister) {
       store.enableStoreEventNotifications(storeListener);
     }
+    if (eventThreadLocal instanceof NoOpEventThreadLocalImpl) {
```

Comment:
You need to invert the if and the synchronized calls, otherwise changes could have happened before you reach the exclusive block. Or you need to perform the check inside again.